pkg/cosi/provisioner.go

```diff
@@ -0,0 +1,450 @@
+		if nbErr, ok := err.(*nb.RPCError); ok {
+			if nbErr.RPCCode == "BUCKET_ALREADY_EXISTS" {
+				msg := fmt.Sprintf("Bucket %q already exists", r.BucketName)
+				log.Error(msg)
+				return status.Error(codes.InvalidArgument, msg)
+			}
+		}
+		return fmt.Errorf("failed to create bucket %q with error: %v", r.BucketName, err)
+	}
+
+	log.Infof("✅ Successfully created bucket %q", r.BucketName)
+	return nil
+}
+
+// CreateAccount creates a new bar account
+func (r *APIRequest) CreateAccount() (*nb.S3AccessKeys, error) {
+	log := r.Provisioner.Logger
+	_, err := r.SysClient.NBClient.ReadBucketAPI(nb.ReadBucketParams{Name: r.BucketName})
+	if nbErr, ok := err.(*nb.RPCError); ok && nbErr.RPCCode == "NO_SUCH_BUCKET" {
+		return nil, status.Error(codes.Internal, nbErr.RPCCode)
+	}
+	accountInfo, err := r.SysClient.NBClient.CreateAccountAPI(nb.CreateAccountParams{
+		Name:              r.AccountName,
+		Email:             r.AccountName,
+		DefaultResource:   "",
+		HasLogin:          false,
+		S3Access:          true,
+		AllowBucketCreate: false,
+		BucketClaimOwner:  r.BucketName,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	var accessKeys nb.S3AccessKeys
+	// if we didn't get the access keys in the create_account reply we might be talking to an older noobaa version (prior to 5.1)
+	// in that case try to get it using read account
+	if len(accountInfo.AccessKeys) == 0 {
+		log.Info("CreateAccountAPI did not return access keys. calling ReadAccountAPI to get keys..")
+		readAccountReply, err := r.SysClient.NBClient.ReadAccountAPI(nb.ReadAccountParams{Email: r.AccountName})
+		if err != nil {
+			return nil, err
+		}
+		accessKeys = readAccountReply.AccessKeys[0]
+	} else {
+		accessKeys = accountInfo.AccessKeys[0]
+	}
+
+	log.Infof("✅ Successfully created account %q with access to bucket %q", r.AccountName, r.BucketName)
+	return &accessKeys, nil
+}
+
+// DeleteAccount deletes the bar account
+func (r *APIRequest) DeleteAccount() error {
+
+	log := r.Provisioner.Logger
+	log.Infof("deleting account %q", r.AccountName)
+	err := r.SysClient.NBClient.DeleteAccountAPI(nb.DeleteAccountParams{Email: r.AccountName})
+
+	if err != nil {
+		if nbErr, ok := err.(*nb.RPCError); ok && nbErr.RPCCode == "NO_SUCH_ACCOUNT" {
+			log.Warnf("Account to delete was not found %q", r.AccountName)
+		} else {
+			return fmt.Errorf("failed to delete account %q. got error: %v", r.AccountName, err)
+		}
+	} else {
+		log.Infof("✅ Successfully deleted account %q", r.AccountName)
+	}
+
+	return nil
+}
+
+// DeleteBucket deletes the cosi bucket **including data**
+func (r *APIRequest) DeleteBucket() error {
+	// TODO we should support the different delete options - not completely closed by the KEP
+	var err error
+	log := r.Provisioner.Logger
+	info, err := r.SysClient.NBClient.ReadBucketAPI(nb.ReadBucketParams{Name: r.BucketName})
+	if nbErr, ok := err.(*nb.RPCError); ok && nbErr.RPCCode != "NO_SUCH_BUCKET" {
+		log.Warnf("Bucket to delete was not found %q", r.BucketName)
```

Comment:
one doubt, in condition we are checking the error code not `NO_SUCH_BUCKET`, and in warning bucket not found, So `if` block only executes when error code is `NO_SUCH_BUCKET`